### PR TITLE
The longswipe special attack can't go through walls

### PR DESCRIPTION
**Pull request by sethsimon**
Fixes [#16933](https://github.com/CleverRaven/Cataclysm-DDA/issues/16933).

Adds a check to `mattack::longswipe` so that if there's something impassable between the monster and the target, it bounces off of the [terrain name].

### CODE REVIEW

File: src/monattack.cpp [hunk 1]
```diff
@@ -3750,12 +3750,28 @@ bool mattack::longswipe(monster *z)
     if( target == nullptr ) {
         return false;
     }
+    if( rl_dist( z->pos(), target->pos() ) > 3 || !z->sees( *target ) ) {
+        return false; //out of range
+    }
+    //Is there something impassable blocking the claw?
+    std::vector<tripoint> line = g->m.find_clear_path( z->pos(), target->pos() );
```
Comment (codemime) [inline src/monattack.cpp]: Can be `const` or inlined:
```c++
for( const auto &pnt : g->m.find_clear_path( z->pos(), target->pos() ) ){
```

File: src/monattack.cpp [hunk 2]
```diff
@@ -3750,12 +3750,28 @@ bool mattack::longswipe(monster *z)
...
+    for( auto &i : line ){
+        ter_t terrain = g->m.ter_at( i );
```
Comment (codemime) [inline src/monattack.cpp]: Can be `const` too.

File: src/monattack.cpp [hunk 3]
```diff
@@ -3750,12 +3750,28 @@ bool mattack::longswipe(monster *z)
...
+        if( terrain.movecost == 0 ){
+            //If we're here, it's an unadjacent attack, which is only attempted 1/5 of the time.
+            if( !one_in( 5 ) ) {
+                return false;
+            }
+            target->add_msg_player_or_npc( _( "The %1$s thrusts a claw at you, but it bounces off the %2$s!" ),
+                                           _( "The %1$s thrusts a claw at <npcname>, but it bounces off the %2$s!" ),
+                                           z->name().c_str(), terrain.name.c_str() );
+            z->moves -= 150;
```
Comment (codemime) [inline src/monattack.cpp]: `z->mod_moves( -150 )` would be better.

File: src/monattack.cpp (hunk 1 above)
Comment (sethsimon) [inline src/monattack.cpp]: Noted, thanks.

File: src/monattack.cpp (hunk 2 above)
Comment (sethsimon) [inline src/monattack.cpp]: Noted.

File: src/monattack.cpp (hunk 3 above)
Comment (sethsimon) [inline src/monattack.cpp]: Noted, thanks. Unfortunately, I can't work on this till tomorrow, then I'll make those changes.

File: src/monattack.cpp [hunk 4]
```diff
@@ -3750,12 +3750,28 @@ bool mattack::longswipe(monster *z)
...
         return false;
     }
+    if( rl_dist( z->pos(), target->pos() ) > 3 || !z->sees( *target ) ) {
+        return false; //out of range
+    }
+    //Is there something impassable blocking the claw?
+    std::vector<tripoint> line = g->m.find_clear_path( z->pos(), target->pos() );
+    for( auto &i : line ){
+        ter_t terrain = g->m.ter_at( i );
+        if( terrain.movecost == 0 ){
```
Comment (Coolthulhu) [inline src/monattack.cpp]: There is `map::impassable` for that. Checking only the terrain will skip furniture and vehicle boards.

Comment (sethsimon) [inline src/monattack.cpp]: Good catch, thanks!

**Comment by sethsimon** [comment]
Is there a good way to say "The claw bounces off the car's quarterpanel" instead of the more silly-sounding "The claw bounces off of the dirt?" I didn't see anything in `map.cpp`. Maybe `map::disp_name` needs an update.

**Comment by Coolthulhu** [comment]
Currently no way other than checking everything.
You could add a relevant function to `map` (better not reuse `disp_name` since it is used in multiple places and some may assume it ignores vehicles) or for now just drop the obstacle naming and have it bounce off something unspecified.

**Comment by sethsimon** [comment]
Alright, I'll remind myself to do that sometime this week. I like the idea of adding a default `bool include_vehicles = false` argument since `disp_name` is a pretty good name for such a function (and `disp_name_include_vehicles` is kind of wordy).

**Comment by illi-kun** [comment]
> Is there a good way to say "The claw bounces off the car's quarterpanel" instead of the more silly-sounding "The claw bounces off of the dirt?"

My version of fixing this is available in #16279, and if you will implement more general solution you can apply it to the place changed in the mentioned PR.

**Comment by Coolthulhu** [comment]
Relatively minor: the longswipe will always result in bouncing off, but only 1 in 5 chance of actually attacking if it can't bounce off.
That is, critters with longswipe are much more likely to use their ranged attack if it can't hit.

**Pull request merged**